lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -122,6 +130,100 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
+  return getSpecializedType();
+}
+
+bool ReabstractionInfo::isFullSpecialization() const {
+  return !hasArchetypes(getOriginalParamSubstitutions());
+}
+
+bool ReabstractionInfo::isPartialSpecialization() const {
+  return hasArchetypes(getOriginalParamSubstitutions());
+}
+
+void ReabstractionInfo::createSubstitutedAndSpecializedTypes() {
+  auto &M = OriginalF->getModule();
+
+  // Find out how the function type looks like after applying the provided
+  // substitutions.
+  if (!SubstitutedType) {
+    SubstitutedType = createSubstitutedType(
+        OriginalF, CallerInterfaceSubs, HasUnboundGenericParams);
+  }
+  assert(!SubstitutedType->hasArchetype() &&
+         "Substituted function type should not contain archetypes");
+
+  // Check which parameters and results can be converted from
+  // indirect to direct ones.
+  NumResults = SubstitutedType->getNumIndirectResults();
+  Conversions.resize(NumResults + SubstitutedType->getParameters().size());
+
+  CanGenericSignature CanSig;
+  if (SpecializedGenericSig)
+    CanSig = SpecializedGenericSig->getCanonicalSignature();
+  Lowering::GenericContextScope GenericScope(M.Types, CanSig);
+
+  if (SubstitutedType->getNumDirectResults() == 0) {
+    // The original function has no direct result yet. Try to convert the first
+    // indirect result to a direct result.
+    // TODO: We could also convert multiple indirect results by returning a
+    // tuple type and created tuple_extract instructions at the call site.
+    unsigned IdxForResult = 0;
+    for (SILResultInfo RI : SubstitutedType->getIndirectResults()) {
+      assert(RI.isIndirect());
+      if (RI.getSILType().isLoadable(M) && !RI.getType()->isVoid()) {
+        Conversions.set(IdxForResult);
+        break;
+      }
+      ++IdxForResult;
+    }
+  }
+
+  // Try to convert indirect incoming parameters to direct parameters.
+  unsigned IdxForParam = NumResults;
+  for (SILParameterInfo PI : SubstitutedType->getParameters()) {
+    if (PI.getSILType().isLoadable(M) &&
+        PI.getConvention() == ParameterConvention::Indirect_In) {
+      Conversions.set(IdxForParam);
+    }
+    ++IdxForParam;
+  }
+
+  // Produce a specialized type, which is the substituted type with
+  // the parameters/results passing conventions adjusted according
+  // to the converions selected above.
+  SpecializedType = createSpecializedType(SubstitutedType, M);
+}
+
+/// Create a new substituted type with the updated signature.
+CanSILFunctionType
+ReabstractionInfo::createSubstitutedType(SILFunction *OrigF,
+                                         const SubstitutionMap &SubstMap,
+                                         bool HasUnboundGenericParams) {
+  auto &M = OrigF->getModule();
+  auto OrigFnTy = OrigF->getLoweredFunctionType();
+
+  // First substitute concrete types into the existing function type.
+  auto FnTy = OrigFnTy->substGenericArgs(
+      M, QueryTypeSubstitutionMap{SubstMap.getMap()},
```

Comment:
That's odd. Really the `ArrayRef<Substitution>` version should be removed in favor of a SubstitutionMap one. I think it should be added back. :-)